corehq/form_processor/models/cases.py

```diff
@@ -334,11 +334,13 @@ def hard_delete_cases_before_cutoff(self, cutoff, dry_run=True):
                 case_tombstones = [DeletedSQLDoc(doc_id=case.case_id, object_class_path=class_path,
                                                  domain=case.domain, deleted_on=case.deleted_on)
                                    for case in queryset]
```

Comment:
Iterating over `queryset` (which loads the full result set into memory) is the thing that would significant memory pressure, not the `.delete()` call below. A possibly more efficient technique would be to run the query in a loop and limit the number of results. Something like this
```py
while cases := queryset[:2000]:
    chunk = [DeletedSQLDoc(doc_id=case.case_id, object_class_path=class_path,
                             domain=case.domain, deleted_on=case.deleted_on)
             for case in cases]
    tombstone_count = len(DeletedSQLDoc.objects.bulk_create(chunk, ignore_conflicts=True))
    counts['tombstone'] += tombstone_count
    chunked_queryset = self.using(db_name).filter(case_id__in=[doc.doc_id for doc in chunk])
    deleted_counts = chunked_queryset.delete()[1]
    for obj_class, count in deleted_counts.items():
        counts[obj_class] += count
```
Aside: can't believe I'm recommending the walrus operator, but I think this is the one place where it's most justified.

Reply:
thanks for the suggestion! Just curious though, the django docs has this line:
> You cannot call delete() on a [QuerySet](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#django.db.models.query.QuerySet) that has had a slice taken or can otherwise no longer be filtered.

Does this apply here to the chunked queryset? or am I just misunderstanding this line?

Reply:
ohhh nevermind I misread the code change